ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/StateTransition.java

```diff
@@ -115,16 +113,21 @@ public BeaconState processSlots(BeaconState preState, UInt64 slot)
           slot);
       BeaconState state = preState;
       while (state.getSlot().compareTo(slot) < 0) {
-        state = processSlot(state);
+        final UInt64 currentSlot = state.getSlot();
+        final SpecVersion spec = specProvider.getSpec(currentSlot);
+        state = processSlot(spec, state);
         // Process epoch on the start slot of the next epoch
-        if (state
-            .getSlot()
-            .plus(UInt64.ONE)
-            .mod(specConfig.getSlotsPerEpoch())
-            .equals(UInt64.ZERO)) {
-          state = epochProcessor.processEpoch(state);
+        if (state.getSlot().plus(UInt64.ONE).mod(spec.getSlotsPerEpoch()).equals(UInt64.ZERO)) {
+          state = spec.getEpochProcessor().processEpoch(state);
         }
         state = state.updated(s -> s.setSlot(s.getSlot().plus(UInt64.ONE)));
+
+        // Upgrade state if necessary
+        final Optional<StateUpgrade<?>> stateUpgrade =
+            stateUpgradeProvider.getStateUpgrade(state.getSlot());
+        if (stateUpgrade.isPresent()) {
+          state = stateUpgrade.get().upgrade(state);
+        }
```

Comment:
Given upgrades only happen on epoch boundaries, it might be better to only perform this check inside the if block for processing epochs.  That way we do 32 times fewer map lookups.  Most of the time it won't matter but during non-finalisation we can wind up processing a lot of empty slots.